### PR TITLE
Support zfs raid levels during install

### PR DESCRIPTION
**Pull request by zedi-pramodh**
Provides grub parameters to explicitly install zfs and pick the raid level.
Support for single disk install of ZFS is added.

Testing done
=============
1) Single disk install grub parameters : eve_install_zfs eve_zfs_raid_level=none eve_install_disk=sda eve_persist_disk=sdb
2) Mirror install grub parameters : eve_install_zfs eve_zfs_raid_level=mirror eve_install_disk=sda eve_persist_disk=sdb,sdc
3) Raidz1 install grub parameters : eve_install_zfs eve_zfs_raid_level=raid5 eve_install_disk=sda eve_persist_disk=sdb,sdc,sdd
4) Raidz2 install grub parameters : eve_install_zfs eve_zfs_raid_level=raid6 eve_install_disk=sda eve_persist_disk=sdb,sdc,sdd,sde
5) Verified that ext4 installs if eve_install_zfs is not provided.

Signed-off-by: Pramodh Pallapothu <pramodh@zededa.com>

### CODE REVIEW

**Comment by eriknordmark** [comment]
In the API/UI are we naming the terms "mirror", "raid5", etc? Or `raidz1"` etc?

It makes sense for us to use consistent user-visible terms for these.

**Comment by eriknordmark** [comment]
> * Single disk install grub parameters : eve_install_zfs eve_zfs_raid_level=none eve_install_disk=sda eve_persist_disk=sdb

Is there also a way to run with a single install+persist disk and zfs?

File: docs/BOOTING.md [hunk 1]
```diff
@@ -127,6 +127,8 @@ in grub.cfg with graphical GRUB menu to get the device to boot again.
     10. `eve_install_skip_persist` - do not install persist partition onto device. May be selected from graphical GRUB menu.
     11. `eve_install_skip_rootfs` - do not install rootfs partition onto device. May be selected from graphical GRUB menu.
     12. `eve_install_skip_zfs_checks` - install zfs by skipping minimum requirement checks.
+    13. `eve_install_zfs` - always install zfs, will not install ext4 if zfs install fails.
+    14. `eve_zfs_raid_level` - Sets raid level for zfs storage. Valid values are none,mirror,raid5,raid6. Default value is none.
```
Comment (giggsoff) [inline docs/BOOTING.md:131]: Do we need `eve_install_zfs`? And from code I can see that we parse `eve_zfs_raid_level` only if `eve_install_zfs` defined. Why?

Comment (eriknordmark) [reply docs/BOOTING.md:131]: You mean if eve_zfs_raid_level is set at all it means we should use zfs? That makes sense.

File: pkg/mkimage-raw-efi/install [hunk 2]
```diff
@@ -265,11 +268,21 @@ DISK_WITH_P3=""
 DISKS_TO_MERGE_COUNT=0
 MULTIPLE_DISKS=false
```
Comment (eriknordmark) [inline pkg/mkimage-raw-efi/install:269]: Are these variables still being used or can we delete them entirely?

File: docs/BOOTING.md (hunk 1 above)
Comment (zedi-pramodh) [reply docs/BOOTING.md:131]: "Is there also a way to run with a single install+persist disk and zfs?"

Yes it works and I tested it. I also tested raid none and multiple disks.

"In the API/UI are we naming the terms "mirror", "raid5", etc? Or raidz1" etc?

It makes sense for us to use consistent user-visible terms for these."

In UI we display raid5, raid6, raid1, so I will change mirror to raid1

File: pkg/mkimage-raw-efi/install (hunk 2 above)
Comment (zedi-pramodh) [reply pkg/mkimage-raw-efi/install:269]: They are used in following cases:
MULTIPLE_DISKS is really used for logging to installer.log where we will know if persist was installed on multiple disks.
DISKS_TO_MERGE_COUNT is useful to check zfs minimum requirements checks. Say install was performed just provided multiple disks (like old way of doing zfs installs) but without explicit eve_install_zfs, then we need to check for disks count.

File: docs/BOOTING.md (hunk 1 above)
Comment (zedi-pramodh) [reply docs/BOOTING.md:131]: Yeah I agree we do not need additional eve_install_zfs, I will remove that.

**Comment by zedi-pramodh** [comment]
Updated commit with review comments.

File: docs/BOOTING.md [hunk 3]
```diff
@@ -127,6 +127,7 @@ in grub.cfg with graphical GRUB menu to get the device to boot again.
     10. `eve_install_skip_persist` - do not install persist partition onto device. May be selected from graphical GRUB menu.
     11. `eve_install_skip_rootfs` - do not install rootfs partition onto device. May be selected from graphical GRUB menu.
     12. `eve_install_skip_zfs_checks` - install zfs by skipping minimum requirement checks.
+    13. `eve_install_zfs_with_raid_level` - Sets raid level for zfs storage. Valid values are none,raid1,raid5,raid6. Default value is none.
```
Comment (eriknordmark) [inline docs/BOOTING.md:130]: My earlier question seems to have been missed so after invoking grep I'll make it into a statement.
We currently do not use raid5 nor raid6 in the API nor in the EVE code, but we do use raid1, raidz1, raidz2.

It doesn't make sense introducing raid5 and raid6 in the installer; we should at least use consistent terms.

Comment (zedi-pramodh) [reply docs/BOOTING.md:130]: Actually we display None, RAID1, RAID5 and RAID6 on UI today. We made that decision as per your suggestion in one of our storage team calls because raidz2 and raidz1 are zfs specific and we wanted to display industry standard terms, that also helps in future if we decide to use some other filesystem. That's why I made installer to be in sync with what user sees in UI.

Comment (eriknordmark) [reply docs/BOOTING.md:130]: But that's already inconsistent because the API calls them raidz1 raidz2. 
Please make it consistent one way or another.

Note that we can change the enum strings in the API without any wire compatibility issues; the protobuf wire encoding just uses the tags and the types.
So if we want to use Raid5,6,7 then change the API and EVE to use those and not just the UI. Also the zmsg API in zedcloud should match.

Comment (zedi-pramodh) [reply docs/BOOTING.md:130]: Working on that.

**Comment by zedi-pramodh** [comment]
Updated the commit by changing EVE API to reflect the raid names to be in sync with UI and installer.
Tested the new installation and verified that raid names are in sync between installer, EVE and UI

**Review by eriknordmark — APPROVED** [review]
LGTM
